text/0000-array-private-classes

```diff
@@ -0,0 +1,148 @@
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of [hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding) but offer a general and stable interface. Returning interfaces instead of concrete classes allows changing the implementation. Usually, one must return the most general type that fullfils the contract of the function (in the case of the functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more general. Iterators implementation details are not public. Internal classes used by implementation like `*Keys`, `*Values` and `*Pairs` are now [opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally, when using these collection classes, clients are not interested by the iterators implementation, but by the types these iterators return and that is provided by the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to understand for clients of collection classes.
+- Makes the standard library more simple by hiding 18 specialised classes from stdlib of which 3 are from `builtin`.
+- The interface `RewindableIterator` is added to create rewindable iterators (can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on Github shows that the impact is very limited.
+
+# Detailed design
+
+Iterating functions in collections `keys`, `values` and `pairs` are changed to return `Iterator` and the classes that implement these iterators are made private. Here are the full implementation of these functions for the `Array` class (changes in other collection classes are identical).
+
+```pony
+  fun keys(): Iterator[USize]^ =>
+    """
+    Return an iterator over the indices in the array.
+    """
+    _ArrayKeys[A, this->Array[A]](this)
+
+  fun values(): RewindableIterator[this->A]^ =>
+    """
+    Return an iterator over the values in the array.
+    """
+    _ArrayValues[A, this->Array[A]](this)
+
+  fun pairs(): Iterator[(USize, this->A)]^ =>
+    """
+    Return an iterator over the (index, value) pairs in the array.
+    """
+    _ArrayPairs[A, this->Array[A]](this)
+```
+
+As the function `values` of class `Array` uses an iterator with a `rewind` function that is not part of the `Iterator` interface, a new interface `RewindableIterator` is added to enable creation of rewindable iterators.
+
+Note: To remain consistent with `Array` behaviour, functions `keys` and `pairs` should return a `RewindableIterator` too but we limited the API change to minimum as we did not understood why it was not already the case.
+
+```pony
+interface RewindableIterator[A] is Iterator[A]
```

Comment:
There's no need for `is Iterator[A]` as far as I can tell here. Is this an attempt to make sure in the future that RewindableIterator doesn't diverge at all from Iterator?

Reply:
That's true for `interface`but I just kept with the coding style of the stdlib where numerous classes (i.e. `Array`, `String`, `List`, `Range`, `Reverse`, `HashSet`, etc.)  explicitly name the interface they implement.

Reply:
So in the case of those, it made sure that if the Iterator interface was changed that all those concrete classes would fail compilation, is that the intention here? If yes, I think there is a larger discussion about the idea of duplicating the methods from Iterator in RewindableIterator and then using `is` to enforce the Iterator interface. 

It feels odd. It might be worth it to avoid having an interface that is only `rewind` and the returning the two interfaces, but, it does feel odd and I think the reasoning is worth discussion.

I know you had some conversation around this with @jemc. I think you should discuss with him some supporting justification for this. Is this a one off? Is this establishing a pattern that we want to use in the standard library? Is it a pattern that is already being used in the standard library?

Reply:
My suggestion would be to keep the `is Iterator[A]` part and remove the explicit `next` and `has_next` definitions.

For an `interface`, that roughly reads as:
- "To implement `RewindableIterator[A]`, you need to implement everything `Iterator[A]` requires (which doesn't need to be repeated here), but you additionally need this `rewind` method."

Or from the perspective of the interface user:
- "If you get a `RewindableIterator[A]`, it can do everything `Iterator[A]` can do (which doesn't need to be repeated here), but you additionally can use it to `rewind`"

Reply:
As noted below by me, I prefer the alternative design of not having `RewindableIterator[A]` overlap with `Iterator[A]` and instead have `Iterator[A] & Rewindable[A]` define `RewindableIterator[A]`.

Reply:
I don't understand what is finally desired here. What would be the `RewindableIterator` (or `CollectionIterator`) definition so that I can include it in the RFC text?